### PR TITLE
Handle bindings with leading :

### PR DESCRIPTION
**Pull request by Argn0**
closes https://github.com/qutebrowser/qutebrowser/issues/3204

---
This change is [Reviewable](https://reviewable.io/reviews/qutebrowser/qutebrowser/3528)

### CODE REVIEW

**Comment by Argn0** [comment]
Executing commands with extra semicolons like `::::::open github.com` work because they are lstripped before they are passed to the parser. 

`    def _parse_all_gen(self, text, *args, aliases=True, **kwargs):
        """Split a command on ;; and parse all parts.

        If the first command in the commandline is a non-split one, it only
        returns that.

        Args:
            text: Text to parse.
            aliases: Whether to handle aliases.
            *args/**kwargs: Passed to parse().

        Yields:
            ParseResult tuples.
        """
        text = text.strip().lstrip(':').strip()
`

**Comment by The-Compiler** [comment]
This still doesn't normalize things like `: nop` which work fine, and wouldn't normalize them for other places getting a command for a binding (though I don't think there are any at the moment).

What do you think about just doing the same normalizing (`text = text.strip().lstrip(':').strip()`) in `CommandParser.parse`? This would also mean that running `:message-info foo ;; :message-info bar` would work properly, which seems a somewhat common mistake as well.

**Comment by Argn0** [comment]
Whoops it seems like I broke a lot of tests. I'll report back when I have fixed those.

File: qutebrowser/commands/runners.py [hunk 1]
```diff
@@ -195,6 +196,8 @@ def parse(self, text, *, fallback=False, keep=False):
             return ParseResult(cmd=None, args=None, cmdline=cmdline)
 
         args = self._split_args(cmd, argstr, keep)
+        args = [x.strip().lstrip(':').strip() if x is not ':' and x is not ' '
+                else x for x in args]
```
Comment (The-Compiler) [inline qutebrowser/commands/runners.py]: I don't think it makes sense to apply this on the arguments.

Comment (Argn0) [reply qutebrowser/commands/runners.py]: Hmm...why not?

**Comment by The-Compiler** [comment]
This still seems like a bad idea. If I do, dunno, `:message-info :foo`, the `:` shouldn't be stripped. It really should never be stripped, except at the beginning of a command.

**Comment by rcorre** [comment]
Just my 2 cents: a config should not be forgiving. If the user inputs something that isn't precise, just error and let them fix it. Adding special cases to guess the intent of incorrect inputs is a rabbit-hole.

**Comment by The-Compiler** [comment]
Fair point. The reason the stripping was added originally was to make an accidental `::foo` work when e.g. pasting `:foo` in the commandline. vim does this as well, and it probably makes sense to do it.

But for bindings, I guess it's indeed fine to just reject a binding with an extra `:` in the config. Certainly easier to reason about than what we're doing here. Probably that extra stripping should really only occur for the commandline, not anywhere else.

@Argn0 is this something you'd like to work on, or should we close this?

**Comment by Argn0** [comment]
Wouldn't disallowing to use bindings with extra semicolons be very unintuitive and break already existing bindings?

**Comment by Argn0** [comment]
As far as I can tell :bind only binds key macros and not really commands.

Would it make sense to allow the user to

**Comment by The-Compiler** [comment]
I don't think it's unintuitive. In theory, qutebrowser should use `:foo` everywhere the commandline is involved (typing `:foo` and using `:set-cmd-text`) and use commands without `:` everywhere else (bindings, `:later`, `:repeat`, etc.).

That actually was the case until #2488 (cc @Carpetsmoker) and probably until 64b783d9c056045bcb5caa3d9ab8543b54fc4e18 with `:bind`. However, this was unintentional - that PR intended to only affect things like entering `::save` in the commandline (e.g. by copy-pasting) but also affected all other places running a command.

Seeing in how many corner cases we're getting into even with this change, I agree it'd be better to be strict again. It's probably worth the breakage - with `:bind` it's quite hard to do a wrong binding accidentally (as you get completions for commands), and with a `config.py` people signed up for breakage on updates anyways :wink:

**Pull request closed without merging**